src/tree.ts

```
@@ -0,0 +1,561 @@
+    return new Fragment(this.ends[L] as MQNode, this.ends[R] as MQNode);
+  };
+
+  eachChild (yield_:(el:MQNode) => boolean | undefined) {
+    eachNode(this.ends, yield_);
+    return this;
+  };
+
+  foldChildren <T>(fold:T, yield_:(fold:T, el:MQNode) => T) {
+    return foldNodes(this.ends, fold, yield_);
+  };
+
+  withDirAdopt (dir:Direction, parent:MQNode, withDir:MQNode, oppDir:MQNode) {
+    const self = this.getSelfNode();
+    new Fragment(self, self).withDirAdopt(dir, parent, withDir, oppDir);
+    return this;
+  };
+
+  adopt (parent:MQNode, leftward:NodeRef, rightward:NodeRef) {
+    var self = this.getSelfNode();
+    new Fragment(self, self).adopt(parent, leftward, rightward);
+    return this.getSelfNode();
+  };
+
+  disown () {
+    var self = this.getSelfNode();
+    new Fragment(self, self).disown();
+    return this;
+  };
+
+  remove () {
+    this.jQ.remove();
+    return this.disown();
+  };
+
+  isParentSimpleSubscript () {
+    if (!this.parent) return false;
+    if (!(this.parent.parent instanceof SupSub)) return false;
+
+    // Mathquill is gross. There are many different paths that
+    // create subscripts and sometimes we don't even construct
+    // true instances of `LatexCmds._`. Another problem is that
+    // the relationship between the sub and the SupSub isn't
+    // completely setup during a paste at the time we check
+    // this. I wanted to use: `this.parent.parent.sub !== this.parent`
+    // but that check doesn't always work. This seems to be the only
+    // check that always works. I'd rather live with this than try
+    // to change the init order of things.
+    if (!this.parent.jQ.hasClass('mq-sub')) return false;
+
+    return true;
+  };
+
+  getSelfNode () {
+    // dumb dance to tell typescript that we eventually become a MQNod
+    return this as any as MQNode
+  }
+
+  // Overridden by child classes
+  parser ():Parser<MQNode> { return undefined as any }; // TODO - is Parser<MQNode> correct?
```

For these method that are actually "abstract", perhaps we should throw instead of returning something weird?

---

I'm slightly worried about them being called and us throwing when we don't need to. I'd vote to leave this in temporarily and think about a cleanup where we are more careful about these abstract methods. I'd prefer for typescript to actually complain if a certain method isn't defined rather than define it and hope we always remember to override it. One thing that comes to mind is to define an interface and make sure each usable Node type implements it.